### ISSUE TITLE
Some confusions about the code

### ISSUE THREAD

**Issue by ganji15** [stevenygd/WAGE.pytorch#1]
Hi, @stevenygd. I have run your code and find that your implementation is very helpful. However, I have serval confusions about the code as follows:
1. I find the optimization is performed on model.weight_acc, which may be the copy of the original weights. But why not directly update the original weights such as model.named_paramerters()?  What should I do if I want to update the network's  weights with the optimizer like Adam?https://github.com/stevenygd/WAGE.pytorch/blob/7fc7ef985516235fe114b8a320a1c0b217d61883/utils.py#L70-L72 

2. the function in utils.train(), wage_grad_clip, is corresponding to the grad_quantizer.  If we need to clip weight, why use models.QG instead of models.QW? I noticed that model.QG introduces the uniform noisy.
https://github.com/stevenygd/WAGE.pytorch/blob/7fc7ef985516235fe114b8a320a1c0b217d61883/utils.py#L70
https://github.com/stevenygd/WAGE.pytorch/blob/7fc7ef985516235fe114b8a320a1c0b217d61883/train.py#L197-L205

3. I also found that you remove the quantization of the input. However, in the original implementation, the input is quantized.  I wonder whether it empirically/theoretically will give a better result if the input quantization is removed?

4. I also test the MNIST dataset, and I found the result is slightly worse than that in the paper (i.e. 99.36% vs.  99.60%). Moreover, I find it can only achieve 99.40% with LeNet5 on MNIST dataset when using the PyTorch platform. I wonder whether this relatively low-precision of LeNet5 is caused by the PyTorch platform?

I will greatly appreciate your kind helps and valuable opinions. Thanks.

**Comment by stevenygd** [comment]
Thanks for your interest in the code base! Responding to your questions:
1. `model.weight_acc` is the 8-bit weight accumulator we kept during the training. Note that WAGE kept such accumulator during training, but use the 2-bit weights (which is furtherly quantized from`weight_acc`) during inference. As a result, weight updates will be accumulated in `weight_acc`.
2. If I remembered in WAGE, only the gradients are quantized with stochastic rounding. That's the reason for using QG, which uses stochastic rounding.
3. In some sense, the input is already quantized into 8 bits (since RGB images are 0-255, which is represented in 8 bits). Adding quantization of the inputs might benefit since it re-scale the input to an appropriate scale.
4. The codes are written to reproduce the performance in CIFAR. WAGE is quite sensitive to hyper-parameter so if you change the network architecture, you might need to re-tune the hyper-parameter carefully.

Hope this help.
